api/server/routecache/cacher.go

```diff
@@ -0,0 +1,101 @@
+package routecache
+
+import (
+	"context"
+	"errors"
+	"github.com/iron-io/functions/api/models"
+	"github.com/iron-io/runner/common"
+	"math"
+	"sync"
+)
+
+type Cacher interface {
+	PrimeCache(ctx context.Context)
+	CacheGet(ctx context.Context, appname, path string) (*models.Route, bool)
+	RefreshCache(ctx context.Context, appname string, route *models.Route)
+	ResetCache(ctx context.Context, appname string, delta int)
+}
+
+type DefaultCacher struct {
+	Datastore models.Datastore
+	hotroutes map[string]*Cache
+	mu        sync.Mutex // protects hotroutes
+}
+
+func NewDefaultCacher(ds models.Datastore) (Cacher, error) {
+	if ds == nil {
+		return nil, errors.New("Missing datastore")
+	}
+	return &DefaultCacher{
+		Datastore: ds,
+		hotroutes: make(map[string]*Cache),
+	}, nil
+}
+
+func (c *DefaultCacher) PrimeCache(ctx context.Context) {
```

Comment:
IMHO, this is wrong in many levels. Going from shallow to deep.

Probably, this made in into routescache because of `cacheParetoThreshold`. Otherwise, how would the cache know how to prime itself? If that's the line of reasoning here, then it is an inversion of responsibility. Cache is meant to be used - and not to use (except, of course, for its underlying driver). Caches are not a subset/superset of databases on which you could have some sort of  "cache-level" stored procedure.

However, therein lies a deeper problem: What is this cache for? This cache was created as a way to prevent unnecessary DB conns during HTTP server `/r/` calls. This cache does hold routes - but it has nothing to do with the rest of the request service life cycle - it does not know of DB calls, it does not know of singleflights and etc.

Think about the inversion: if you accept that you can pull some sort of HTTP/Router logic _into_ the `Cacher` interface - you suddenly have two problems: a) it also means you can pull logic into other parts, such as single flight and DB drivers., how would you decide when one starts and other ends? b) how would you handle the situation on which two parts could have competing views of the same pulled logic? 

From what I see, `PrimeCache` should not exist - and should be reverted to the place it was before, operating the now injected `Cacher`.